src/ipips/ipip-0445.md

```diff
@@ -0,0 +1,105 @@
+graph using `raw` codec for it's leaves.
+
+Usefull with unixfs for features like webseeds [#444](https://github.com/ipfs/specs/issues/444).
+
+## Detailed design
+
+The `skip-leaves` CAR Content-Type parameter on :cite[trustless-gateway]
+allows clients to download an entity except blocks with the multicodec
+`raw` (`0x55`).
+
+- When set to `y`, the parameter instructs the gateway not to transmit
+  blocks tagged with the `raw` multicodec.
+- If set to `n`, or left unspecified, the gateway MUST transmit `raw`
+  multicodec blocks.
+
+Importantly, unless explicitly specified as `y`, the default operational
+mode of the gateway MUST assume the value of `skip-leaves` to be `n`.
+
+## Design rationale
+
+### User Benefit
+
+Implementing the `skip-leaves` parameter offers several benefits to users:
+
+1. **Verification Flexibility:** Clients can verify out-of-band (OOB) received
+   files in their deserialized form without necessitating the transmission of
+   raw blocks from the gateway.
+2. **Incremental Download:** Clients can incrementally download files in
+   deserialized forms from non-IPFS servers. Allowing applications to share
+   distribution for IPFS and non IPFS clients.
+3. **Efficient Block Discovery:** With the `skip-leaves` option enabled,
+   clients can quickly discover numerous candidate blocks without being
+   bottlenecked by the gateway's transmission of raw blocks.
+
+### Compatibility
+
+Setting the default value of the `skip-leaves` parameter to `n` ensures
+backward compatibility with existing clients and systems that are unaware
+of this new flag.
+
+### Prevention of Amplification Attacks and Efficient Server Operation
+
+By utilizing the `raw` (`0x55`) codec servers can trivially determine whether
+to fetch or skip a block without having to learn any new information.
+Although more limited and not able to handle unixfs file using dag-pb for their
+leaves, it allows both the client and server to trivially verify a block
+must not be fetched. Preventing issues of Amplification where a server could
+need to fetch multiple orders more data than the client when executing the
+request.
+
+### Why not `dag-scope=skip-leaves` ?
+
+The `dag-scope` parameter determines the overall range of blocks to retrieve,
+while `skip-leaves` selectively filters specific blocks within that range.
+Combining them under one parameter would restrict their combined utility.
+
+For example:
+- A client is streaming a video from a webseed and the user seeked through the
+  video, then the client would send `dag-scope=entity&entity-bytes=42:1337`
+  with `skip-leaves=y` to download the proofs for the required section of the
+  video.
+- A client is verifying an OOB transfered directory in deserialized form,
+  then `dag-scope=all` with  `skip-leaves=y` makes sense.
+
+### Alternatives
+
+An alternative approach would be to request blocks individually.
+However it adds extra round trips and more per HTTP request overhead
+and thus is undesireable.
+
```

Comment:
```suggestion
1. An alternative approach would be to request blocks individually.
However it adds extra round trips and more per HTTP request overhead
and thus is undesireable.

2. An alternative implementation may be to either specify codecs to
include, or codecs to exclude. e.g. `exclude-codec=0x55` or
`include-codec=0x70` to achieve similar results. There may be
broader utility in such an approach, but the use-cases beyond that
proposed above for `skip-leaves` are not clear.
```

:shrug: this _is_ an option, is it better? not sure, I'm not even sure it'd be very different on the implementation level and may afford some other possibilities. It also clears up the "leaves" terminology, which is very unixfs+CIDv1 specific.

Reply:
Intresting idea. I don't think the current API is unixfs specific. I hope `raw` blocks are children-less bytes in all formats that use them.
I guess you could have an ADL that parse the bytes node out of the file using cbor but then you are using IPLD wrong IMO, should really be a cbor cid then.

`exclude-codec` and `include-codec` are interesting however they complexity the server implementation.
I have an idea on how we could combine a small index and a CARv1 DFS body with offloading and zerocopy to make a 400Gbps gateway with extremely low CPU usage implementation and having user input surface complexify the index this situation because I don't just need to map ranges in the unixfs file to ranges in the car, I also need to mark and compute prefix sums for all the codecs in the car.
So given this might anoy me one day and I don't need it, I wouldn't do it unless someone else wants this feature right now.

Reply:
Also one weird thing, if I have a dag that is `dag-json → dag-cbor → raw` and I exclude the `dag-cbor` codec, is the gateway expected to send me 1 or 2 blocks ? Because if the answer is 2 blocks then the gateway has to process every skipped blocks which means the gateway does work which does not reflect on the client. And if the answer is 1 then it does not follow least surprise principle.

Reply:
(bit unprocessed thought) If we are framing this around DAG traversal, `exclude-codec=0x55` could be a sensible attempt at future-proofing, but we can always add it later.

Rationale:

The  `raw` codec provides an interoperable way to act as traversal stop: ensure the child block that was referred will not be used for any further traversal / deserialization, but returned as an opaque blob. Allows for creating non-UnixFS dags that are compatible with existing IPFS ecosystem.

If we add generic  `exclude-codec` then its semantic meaning should also control when the traversal stops: there could be a real world utility for excluding DAG branches based on their root codec, e.g. when they can't be retrieved with usual transports and require special handling. Creating a knob for increasing retrieval interop of data that is partially encoded in proprietary codecs, but now sure how real world need there is for this _right now_.  

One thing that comes to mind is Filecoin which iirc can't be pinned or fetched recursively with standard IPFS tools due to the use of unsupported codecs (cc @hsanjuan @aschmahmann  if i misremembered this, I think we've hit that problem a while ago).

Is there any other use case that could benefit from open-ended `exclude-codec` instead of a flag that skips raw blocks?

Reply:
Ok,  silence suggests we don't have any immediate need for `exclude-codec`.
Let's keep this IPIP limited to `skip-raw-blocks`,  open-ended filtering can be proposed in future IPIP.